examples/test_messaging/test_messaging.c

```diff
@@ -0,0 +1,431 @@
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *     * The name of the author may not be used to endorse or promote
+ *       products derived from this software without specific prior
+ *       written permission.
+ *
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ *   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ *   OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ *   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * test_messaging.c - unit test to ensure NFs can send messages to themselves
+ ********************************************************************/
+
+#include <errno.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <unistd.h>
+
+#include <rte_common.h>
+#include <rte_ip.h>
+#include <rte_mbuf.h>
+#include <rte_mempool.h>
+#include <rte_ring.h>
+#include <rte_malloc.h>
+
+#include "onvm_nflib.h"
+#include "onvm_pkt_helper.h"
+#include "onvm_common.h"
+
+#define NF_TAG "test_messaging"
+#define MAGIC_NUMBER 11
+
+#define TEST_FAILED -1
+#define TEST_INCOMPLETE 0
+#define TEST_PASSED 1
+
+#define TEST_TIME_LIMIT 5
+
+static uint32_t print_delay = 1000000;
+
+struct test_msg_data{
+        int tests_passed;
+        int tests_failed;
+        int test_phase;
+        int test_msg_count;
+        int ring_count;
+        int mempool_count;
+        uint16_t address;
+        struct rte_mempool* msg_pool;
+        struct rte_ring *msg_q;
+        int test_status[3]; 
+        uint64_t start_time;
+};
+
+void
+nf_setup(struct onvm_nf_local_ctx *nf_local_ctx);
+
+void
+nf_msg_handler(void *msg_data, struct onvm_nf_local_ctx *nf_local_ctx);
+
+/*
+ * Frees memory allocated to the test_msg_data struct
+ */ 
+static int
+destroy_test_msg_data(struct test_msg_data **test_msg_data) {
+        if (test_msg_data == NULL) {
+                return 0;
+        }
+        rte_free(*test_msg_data);
+        (*test_msg_data) = NULL;
+        return 0;
+}
+
+/*
+ * Clears the message queues after each test
+ */
+static void
+clear_msg_queue(struct test_msg_data *test_state) {
+        void* msg;
+        while (rte_ring_count(test_state->msg_q) != 0) {
+                rte_ring_dequeue(test_state->msg_q, &msg);
```

Comment:
This is a potential memory leak. We are dequeuing things from the queue, but not freeing either the overall nf message structure or the data inside it.

I think when we call `rte_ring_dequeue(test_state->msg_q)` we get back a `struct onvm_nf_msg`. Inside that struct will be a `void* msg_data` which should be freed with `rte_malloc`.

Then the `onvm_nf_msg` can should be freed by calling `rte_mempool_put(test_state->msg_q, (void *)msg);`